File: 1-1000/0828-count-unique-characters-of-all-substrings-of-a-given-string/0828-count-unique-characters-of-all-substrings-of-a-given-string.cpp
class Solution {
public:
    int uniqueLetterString(string s) {
        vector<int> last(26, -1);
        vector<int> prevLast(26, -1);
        int ans = 0;
        int sum = 0;
        int n = s.size();
        for(int i = 0; i < n; i ++) {
            int c = s[i] - 'A';
            sum += i - last[c];
            sum -= (last[c] - prevLast[c]);
            ans += sum;
            prevLast[c] = last[c];
            last[c] = i;
        }
        return ans;
    }
};

File: 1-1000/0828-count-unique-characters-of-all-substrings-of-a-given-string/0828-count-unique-characters-of-all-substrings-of-a-given-string.java
class Solution {
    public int uniqueLetterString(String s) {
        char[] arr = s.toCharArray();
        int[] last = new int[26];
        Arrays.fill(last, -1);
        int[] secLast = new int[26];
        Arrays.fill(secLast, -1);
        int count = 0;
        int ans = 0;
        for(int i = 0;i < arr.length;i++) {
            int idx = arr[i] - 'A';
            int l = last[idx];
            int sl = secLast[idx];
            count += i - l - l + sl;
            ans += count;
            secLast[idx] = last[idx];
            last[idx] = i;
        }
        return ans;
    }
}